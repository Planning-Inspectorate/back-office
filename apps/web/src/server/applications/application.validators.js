import { createValidator } from '@pins/express';
import { body } from 'express-validator';

export const validateSearchText = createValidator(
	body('SearchApplications')
		.trim()
		.isLength({ min: 1 })
		.withMessage('Enter a search term')
		.isLength({ max: 500 })
		.withMessage('Search term must be 500 characters or fewer')
);
