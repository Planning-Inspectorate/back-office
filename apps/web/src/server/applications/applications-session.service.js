/**
 * @typedef {import('./applications.types').DomainType} DomainType
 * @typedef {import('express-session').Session & { applicationsDomainType?: DomainType, redirectBackTo?: string }} BOASession
 */

/**
 * Set the domainType of the user in the session.
 *
 * @param {BOASession} session
 * @param {DomainType} domainType
 * @returns {void}
 */
export const setSessionApplicationsDomainType = (session, domainType) => {
	session.applicationsDomainType = domainType;
};

/**
 * Get the domainType of the user from the session.
 *
 * @param {BOASession} session
 * @returns {DomainType|null}
 */
export const getSessionApplicationsDomainType = (session) => {
	return session.applicationsDomainType ?? null;
};

/**
 * Set the redirectBackTo for the session.
 *
 * @param {BOASession} session
 * @param {string} redirectBackTo
 * @returns {void}
 */
export const setSessionApplicationsReferBackTo = (session, redirectBackTo) => {
	session.redirectBackTo = redirectBackTo;
};

/**
 * Get the redirectBackTo from the session.
 *
 * @param {BOASession} session
 * @returns {string|null}
 */
export const getSessionApplicationsReferBackTo = (session) => {
	return session.redirectBackTo ?? null;
};

/**
 * Deletes the redirectBackTo from the session.
 *
 * @param {BOASession} session
 */
export const destroySessionApplicationsReferBackTo = (session) => {
	delete session.redirectBackTo;
};
