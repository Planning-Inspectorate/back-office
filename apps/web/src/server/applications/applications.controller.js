import * as applicationsService from './applications.service.js';

/** @typedef {import('@pins/express').ValidationErrors} ValidationErrors */
/** @typedef {import('./applications.locals').ApplicationLocals} ApplicationLocals */
/** @typedef {import('./applications.router').DomainParams} DomainParams */
/** @typedef {import('./applications.types').Application} Application */
/** @typedef {import('./applications.types').DomainType} DomainType */

/**
 * @typedef {object} ViewDashboardRenderProps
 * @property {Application[]} applications
 * @property {Application[]} draftApplications
 */

/**
 * View the domain-specific dashboard for the open applications.
 *
 * @type {import('@pins/express').RenderHandler<ViewDashboardRenderProps,
 * {}, {}, {}, DomainParams>}
 */
export async function viewDashboard({ params }, res) {
	const { domainType } = params;
	const allApplications = await applicationsService.findOpenApplicationsByDomainType(domainType);
	const readyApplications = [];
	const draftApplications = [];

	for (const application of allApplications) {
		if (application.status === 'Draft') {
			draftApplications.push(application);
		} else {
			readyApplications.push(application);
		}
	}

	return res.render('applications/dashboard', {
		applications: readyApplications,
		draftApplications
	});
}

/**
 * @typedef {object} ViewApplicationRenderProps
 * @property {Application} application
 */

/**
 * View the details for a single application.
 *
 * @type {import('@pins/express').RenderHandler<ViewApplicationRenderProps, ApplicationLocals>}
 */
export async function viewApplication({ locals }, response) {
	response.render('applications/application', { application: locals.application });
}
