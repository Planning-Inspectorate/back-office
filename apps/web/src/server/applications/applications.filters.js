import nunjucks from '../app/config/nunjucks.js';

/** @typedef {import('./applications.types').DomainType} DomainType */
/** @typedef {import('nunjucks').Environment} Environment */

/**
 * @typedef {object} urlFilterArguments
 * @property {DomainType=} domainType
 * @property {number=} applicationId
 */

/**
 * Register all the filters under this domain.
 *
 * @type {import('express').RequestHandler<{}, *, *, *, {}>}
 */
export const registerFilters = (req, response, next) => {
	setDisplayValueFilter();
	setUrlFilter();

	next();
};

/**
 * Register the displayValue filter for the domainType
 *
 * @returns {Environment}
 */
const setDisplayValueFilter = () =>
	nunjucks.addFilter('displayValue', (/** @type {DomainType} */ key) => {
		switch (key) {
			case 'case-admin-officer':
				return 'Case admin officer';
			case 'case-officer':
				return 'Case officer';
			case 'inspector':
				return 'Inspector';
			default:
				return '';
		}
	});

/**
 * Register the url filter
 *
 * @returns {Environment}
 */
const setUrlFilter = () =>
	nunjucks.addFilter(
		'url',
		(/** @type {string} */ key, /** @type {urlFilterArguments} */ filterArguments) => {
			const domainUrl = '/applications-service';

			const { domainType, applicationId } = filterArguments || {};

			switch (key) {
				case 'dashboard':
					return `${domainUrl}/${domainType || ''}`;
				case 'create-new-application':
					return `${domainUrl}/create-new-case`;
				case 'view-application':
					return `${domainUrl}/${domainType || ''}/applications/${applicationId || ''}`;
				default:
					return 'app/404';
			}
		}
	);
