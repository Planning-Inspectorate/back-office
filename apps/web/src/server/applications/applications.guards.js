import config from '@pins/web/environment/config.js';
import { assertGroupAccess } from '../app/auth/auth.guards.js';
import { getSessionApplicationsDomainType } from './applications-session.service.js';

/** @typedef {import('./applications.router').DomainParams} DomainParams */
/** @typedef {import('./applications.types').DomainType} DomainType */

/**
 *  Make sure the domainType of the user belongs to the list of possible roles.
 *  OR show 403 page
 *
 *  @type {import('express').RequestHandler<DomainParams>}
 */
export const assertDomainTypeAccess = (req, res, next) => {
	/** @type {Record<DomainType, string>} */
	const domainMap = {
		'case-admin-officer': config.referenceData.applications.caseAdminOfficerGroupId,
		'case-officer': config.referenceData.applications.caseOfficerGroupId,
		inspector: config.referenceData.applications.inspectorGroupId
	};
	const groupId = domainMap[req.params.domainType];

	assertGroupAccess(groupId)(req, res, next);
};

/**
 *  Make sure the domainType of the user is defined
 *  OR redirect to root
 *
 *  @type {import('express').RequestHandler<DomainParams>}
 */
export const assertDomainTypeExists = (req, res, next) => {
	const domainType = getSessionApplicationsDomainType(req.session);

	if (domainType) {
		next();
	} else {
		return res.redirect('/');
	}
};
