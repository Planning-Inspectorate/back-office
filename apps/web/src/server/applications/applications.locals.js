import { getCase } from './common/services/case.service.js';

/** @typedef {import('../app/auth/auth.service').AccountInfo} AccountInfo */

/**
 * @typedef {object} ApplicationsLocals
 * @property {string} serviceName - The name of the service to be displayed in the page header.
 * @property {string} serviceUrl - The root url of the service.
 */

/**
 * Register the non-domain locals for the Applications scope
 *
 * @type {import('express').RequestHandler<*, *, *, *, ApplicationsLocals>}
 */
export const registerLocals = ({ baseUrl }, response, next) => {
	response.locals.serviceName = 'NSIP Applications';
	response.locals.serviceUrl = baseUrl;

	next();
};

/**
 *
 * @param {string[]|null} query
 * @param {boolean} shouldBeDraft
 * @returns {(request: *, response:*, next:*) => void}
 */
export const registerCaseWithQuery = (query, shouldBeDraft = false) => {
	return async (request, response, next) => {
		const { caseId } = response.locals;
		const fullQuery = !query ? null : [...query, 'status'];

		let currentCase;

		try {
			currentCase = caseId
				? await getCase(caseId, fullQuery)
				: { title: '', description: '', status: 'Draft' };
		} catch {
			return throwError(`Failed API request for case ${caseId}`, next);
		}

		const isDraft = currentCase.status === 'Draft';

		if ((isDraft && !shouldBeDraft) || (currentCase.status !== 'Draft' && shouldBeDraft)) {
			return throwError(
				`Trying to load a ${shouldBeDraft ? '' : 'non-'}draft page for a ${
					isDraft ? '' : 'non-'
				}draft case`,
				next
			);
		}

		response.locals.currentCase = currentCase;
		next();
	};
};

// TODO: in case this needs to be used by other locals, move to shared folder
/**
 *
 * @param {string} errorMessage
 * @param {*} next
 * @returns {Promise<void>}
 */
const throwError = (errorMessage, next) => {
	return Promise.resolve()
		.then(() => {
			throw new Error(errorMessage);
		})
		.catch(next);
};
