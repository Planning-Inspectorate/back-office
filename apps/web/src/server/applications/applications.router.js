import { Router as createRouter } from 'express';

const router = createRouter();

export default router;
