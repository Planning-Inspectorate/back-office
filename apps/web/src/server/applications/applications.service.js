import { get } from '../lib/request.js';

/** @typedef {import('./applications.types').Application} Application */
/** @typedef {import('./applications.types').DomainType} DomainType */

/**
 * @param {number} id
 * @returns {Promise<Application>}
 */
export const findApplicationById = (id) => {
	return get(`applications/${id}`);
};

/**
 * @param {DomainType} domainType
 * @returns {Promise<import('./applications.types').Application[]>}
 */
export const findOpenApplicationsByDomainType = (domainType) => {
	return get(`applications/${domainType}`);
};
