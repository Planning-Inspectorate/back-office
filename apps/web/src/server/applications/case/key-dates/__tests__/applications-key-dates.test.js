import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';
import { createTestEnvironment } from '../../../../../../testing/index.js';

import { fixtureCases } from '../../../../../../testing/applications/applications.js';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);

const nocks = () => {
	nock('http://test/').get('/applications/case-team').reply(200, []);
	nock('http://test/').get('/applications/inspector').reply(200, []);

	nock('http://test/').get('/applications/123').times(2).reply(200, fixtureCases[3]);
	nock('http://test/').get('/applications/123').times(2).reply(200, fixtureCases[3]);
};

describe('S51 Advice', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	beforeEach(async () => {
		nocks();
	});

	afterAll(() => {
		nock.cleanAll();
	});

	const baseUrl = '/applications-service/case/123/key-dates';

	describe('Key dates page', () => {
		describe('GET /case/123/key-dates/', () => {
			describe('If user is inspector', () => {
				it('should not render the page', async () => {
					await request.get('/applications-service/inspector');

					const response = await request.get(`${baseUrl}`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toMatchSnapshot();
					expect(element.innerHTML).toContain('problem with your login');
				});
			});

			describe('If user is not inspector', () => {
				it('should render the page', async () => {
					await request.get('/applications-service/case-team');

					const response = await request.get(`${baseUrl}`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toMatchSnapshot();
					expect(element.innerHTML).toContain('Provide details on dates');
				});
			});
		});
	});
});
