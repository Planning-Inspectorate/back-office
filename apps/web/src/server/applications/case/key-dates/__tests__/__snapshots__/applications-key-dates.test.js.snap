// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`S51 Advice Key dates page GET /case/123/key-dates/ If user is inspector should not render the page 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-grid-row\\">
        <div class=\\"govuk-grid-column-two-thirds\\">
            <h1 class=\\"govuk-heading-xl\\">Sorry, there is a problem with your login</h1>
            <p class=\\"govuk-body\\">You are not permitted to access this URL.</p>
            <p class=\\"govuk-body\\"><a href=\\"https://has-appeal.herokuapp.com/help/contact\\" class=\\"govuk-link\\">Contact the Planning Inspectorate Customer Support</a> if
                the problem persists.</p>
        </div>
    </div>
</main>"
`;

exports[`S51 Advice Key dates page GET /case/123/key-dates/ If user is not inspector should render the page 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"top-errors-hook\\" role=\\"alert\\"></div>
    <div class=\\"govuk-grid-row govuk-!-padding-top-9 govuk-!-padding-bottom-2 govuk-!-margin-left-0 govuk-!-margin-bottom-7 pins-masthead-summary\\">
        <div class=\\"govuk-grid-column-full\\">
            <h4 class=\\"govuk-heading-m\\"> CASE/04 </h4>
            <h1 class=\\"govuk-heading-xl\\">Title CASE/04</h1>
            <div class=\\"govuk-grid-row\\">
                <div class=\\"govuk-grid-column-full govuk-body\\">Transport EN, Highways EN
                    <br>
                    <br><strong class=\\"govuk-tag govuk-tag--green\\">Pre-Application</strong>
                </div>
            </div>
        </div>
    </div>
    <div class=\\"govuk-grid-row display--flex\\">
        <div class=\\"pins-list-menu-container\\">
            <aside class=\\"govuk-!-margin-bottom-8 pins-list-menu\\">
                <h2 class=\\"govuk-visually-hidden\\">Case Details Menu</h2>
                <nav>
                    <ul class=\\"gov-list pins-list-menu\\">
                        <li><a class=\\"govuk-link \\" href=\\"/applications-service/case/4/\\"> Overview</a>
                        </li>
                        <li><a class=\\"govuk-link \\" href=\\"/applications-service/case/4/project-information\\"> Project information</a>
                        </li>
                        <li><a class=\\"govuk-link pins-selected-item\\" href=\\"/applications-service/case/4/key-dates\\"> Key dates</a>
                        </li>
                        <li><a class=\\"govuk-link \\" href=\\"/applications-service/case/4/project-team\\"> Project team</a>
                        </li>
                        <li><a class=\\"govuk-link \\" href=\\"/applications-service/case/4/fees\\"> Fees</a>
                        </li>
                        <li><a class=\\"govuk-link \\" href=\\"/applications-service/case/4/project-documentation\\"> Project documentation</a>
                        </li>
                        <li><a class=\\"govuk-link \\" href=\\"/applications-service/case/4/project-updates\\"> Project updates</a>
                        </li>
                        <li><a class=\\"govuk-link \\" href=\\"/applications-service/case/4/examination-timetable\\"> Examination timetable</a>
                        </li>
                    </ul>
                </nav>
            </aside>
        </div>
        <div class=\\"govuk-grid-column-full\\">
            <h3 class=\\"govuk-heading-m\\">Key dates</h3>
            <p class=\\"govuk-body\\">Provide details on dates and deadlines in this project.</p>
            <hr class=\\"govuk-section-break govuk-section-break--l govuk-section-break--visible\\">
            <div class=\\"govuk-accordion\\" data-module=\\"govuk-accordion\\" id=\\"accordion-key-dates\\">
                <div class=\\"govuk-accordion__section\\">
                    <div class=\\"govuk-accordion__section-header\\">
                        <h2 class=\\"govuk-accordion__section-heading\\"><span class=\\"govuk-accordion__section-button\\" id=\\"accordion-key-dates-heading-1\\"><p class='govuk-heading-s'>errors</p></span></h2>
                    </div>
                    <div id=\\"accordion-key-dates-content-1\\" class=\\"govuk-accordion__section-content\\"
                    aria-labelledby=\\"accordion-key-dates-heading-1\\">
                        <div class='pins-key-dates-section'><a class='govuk-link offset-link' href='#'>Manage dates</a>
                            <table class=\\"govuk-table\\">
                                <tbody class=\\"govuk-table__body\\">
                                    <tr class=\\"govuk-table__row\\">
                                        <th scope=\\"row\\" class=\\"govuk-table__header\\">msg</th>
                                        <td class=\\"govuk-table__cell\\"></td>
                                    </tr>
                                </tbody>
                            </table>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</main>"
`;
